src/middlewares/snapshot-middleware.ts

```diff
@@ -20,43 +20,88 @@ export class SnapshotMiddleware {
-                        }
-                    } else if (context.requestHeaders.snapRequest) {
-                        context.returnedExtensions.totalCount = totalCount;
-                    }
-
+                    countPerPage = await this.calculateSnapshotParameters(result, context);
                     startIndex = 0;
                 } else {
                     startIndex = context.snapshotContext.startIndex!;
                     countPerPage = context.snapshotContext.countPerPage!;
                 }
+
+                currentPage = await this.prefetchEntities(
+                    context,
+                    result,
+                    startIndex,
+                    countPerPage,
+                );
+            } else {
+                currentPage = await result.getData(0, undefined);
             }
 
             this.logger.debug('Snapshot middleware finished job', context);
-            return result.getData(startIndex, countPerPage);
+            return currentPage;
         };
     }
+
+    private async calculateSnapshotParameters(result: any, context: PolarisGraphQLContext) {
+        const totalCount = await result.totalCount();
+        let countPerPage = context.requestHeaders.snapPageSize
+            ? Math.min(this.snapshotConfiguration.maxPageSize, context.requestHeaders.snapPageSize)
+            : this.snapshotConfiguration.maxPageSize;
+
+        if (this.snapshotConfiguration.autoSnapshot) {
+            if (totalCount > countPerPage) {
+                context.returnedExtensions.totalCount = totalCount;
+            } else {
+                countPerPage = totalCount;
+            }
+        } else if (context.requestHeaders.snapRequest) {
+            context.returnedExtensions.totalCount = totalCount;
+        }
+
+        return countPerPage;
+    }
+
+    private async prefetchEntities(
+        context: PolarisGraphQLContext,
+        result: any,
+        startIndex: number,
+        countPerPage: number,
+    ) {
+        let prefetchBuffer = context.snapshotContext?.prefetchBuffer || [];
+
+        if (prefetchBuffer.length < countPerPage) {
+            const fetchedData = await this.fetchMoreDataForBuffer(result, startIndex, countPerPage);
+            prefetchBuffer = [...(prefetchBuffer || []), ...(fetchedData || [])];
+        }
+
+        const currentPage = prefetchBuffer.splice(0, countPerPage);
+
+        if (prefetchBuffer.length > 0) {
+            context.returnedExtensions.prefetchBuffer = prefetchBuffer;
+        } else {
+            delete context.returnedExtensions.prefetchBuffer;
+        }
+
+        return currentPage;
+    }
+
+    private fetchMoreDataForBuffer(result: any, startIndex: number, countPerPage: number) {
+        return result.getData(
+            startIndex,
+            countPerPage > this.snapshotConfiguration.entitiesAmountPerFetch
+                ? countPerPage
+                : this.snapshotConfiguration.entitiesAmountPerFetch,
```

Comment:
so if they set max page size bigger than entitiesAmountPerFetch its ok? maybe in constructor check their configuration makes sense